Source/WebCore/html/HTMLAttachmentElement.cpp

```diff
@@ -635,26 +604,78 @@ void HTMLAttachmentElement::updateEnclosingImageWithData(const String& contentTy
     enclosingImage->setAttributeWithoutSynchronization(HTMLNames::srcAttr, AtomString { DOMURL::createObjectURL(document(), Blob::create(&document(), buffer->extractData(), mimeType)) });
 }
 
+void HTMLAttachmentElement::updateImage()
+{
+    if (!m_imageElement)
+        return;
+
+    if (!m_thumbnailImgSrcData.isEmpty()) {
+        m_imageElement->setSrc(AtomString { DOMURL::createObjectURL(document(), Blob::create(&document(), Vector<uint8_t>(m_thumbnailImgSrcData), "image/png"_s)) });
+        return;
+    }
+
+    if (!m_iconImgSrcData.isEmpty()) {
+        m_imageElement->setSrc(AtomString { DOMURL::createObjectURL(document(), Blob::create(&document(), Vector<uint8_t>(m_iconImgSrcData), "image/png"_s)) });
+        return;
+    }
+
+    m_imageElement->setSrc(nullAtom());
+}
+
 void HTMLAttachmentElement::updateThumbnail(const RefPtr<Image>& thumbnail)
 {
+    ASSERT(!isWideLayout());
     m_thumbnail = thumbnail;
-    if (m_innerLegacyAttachment)
-        m_innerLegacyAttachment->updateThumbnail(thumbnail);
     removeAttribute(HTMLNames::progressAttr);
     invalidateRendering();
 }
 
+void HTMLAttachmentElement::updateThumbnailImgSrcData(Vector<uint8_t>&& thumbnailSrcData)
+{
+    ASSERT(isWideLayout());
+    m_thumbnailImgSrcData = WTFMove(thumbnailSrcData);
+    updateImage();
+}
+
 void HTMLAttachmentElement::updateIcon(const RefPtr<Image>& icon, const WebCore::FloatSize& iconSize)
 {
+    ASSERT(!isWideLayout());
     m_icon = icon;
     m_iconSize = iconSize;
-    if (m_innerLegacyAttachment)
-        m_innerLegacyAttachment->updateIcon(icon, iconSize);
     invalidateRendering();
 }
 
+void HTMLAttachmentElement::updateIconImgSrcData(Vector<uint8_t>&& iconSrcData)
+{
+    ASSERT(isWideLayout());
+    m_requestedIconSize = { -1.0, -1.0 };
+    m_iconImgSrcData = WTFMove(iconSrcData);
+    updateImage();
+}
+
+void HTMLAttachmentElement::requestWideLayoutIcon()
+{
+    ASSERT(isWideLayout());
+    if (!document().page() || !document().page()->attachmentElementClient())
+        return;
+    if (m_requestedIconSize.width() < 0.0)
+        return;
```

Comment:
I think you shouldn't need this check if you only call `requestWideLayoutIcon` from `HTMLAttachmentElement` when needed.